src/coreclr/gc/gc.cpp

```diff
@@ -20353,11 +20353,49 @@ void gc_heap::allocate_for_no_gc_after_gc()
     {
         if (current_no_gc_region_info.soh_allocation_size != 0)
         {
+#ifdef USE_REGIONS
+            size_t required = soh_allocation_no_gc;
+            heap_segment* seg = generation_allocation_segment (generation_of (0));
+            size_t available = heap_segment_reserved (seg) - heap_segment_allocated (seg);
+            size_t commit = min (available, required);
+            if (grow_heap_segment (seg, heap_segment_allocated (seg) + commit))
+            {
+                required -= commit;
+                while (required > 0)
+                {
+                    uint8_t* start;
+                    uint8_t* end;
+                    bool allocated_p = global_region_allocator.allocate_basic_region (&start, &end);
```

Comment:
should do `get_new_region` instead of calling `region_allocator` here. all this logic is already in `get_new_region`, and it considers what's already in `free_regions` instead of allocating new regions right away.